src/compiler/compile-props.js

```diff
@@ -278,36 +278,23 @@ export function assertProp (prop, value) {
   }
   var options = prop.options
   var type = options.type
-  var valid = true
-  var expectedType
+  var valid
+  var expectedTypes = []
   if (type) {
-    if (type === String) {
-      expectedType = 'string'
-      valid = typeof value === expectedType
-    } else if (type === Number) {
-      expectedType = 'number'
-      valid = typeof value === 'number'
-    } else if (type === Boolean) {
-      expectedType = 'boolean'
-      valid = typeof value === 'boolean'
-    } else if (type === Function) {
-      expectedType = 'function'
-      valid = typeof value === 'function'
-    } else if (type === Object) {
-      expectedType = 'object'
-      valid = isPlainObject(value)
-    } else if (type === Array) {
-      expectedType = 'array'
-      valid = isArray(value)
-    } else {
-      valid = value instanceof type
+    if (!isArray(type)) {
+      type = [ type ]
+    }
+    for (var i = 0; i < type.length && !valid; i++) {
+      var assertedType = assertType(value, type[i])
+      expectedTypes.push(assertedType.expectedType)
+      valid = assertedType.valid
     }
   }
-  if (!valid) {
+  if (typeof valid !== 'undefined' && !valid) {
```

Comment:
I've just noticed that this was also changed with this [merge](https://github.com/vuejs/vue/commit/e831412f2a4fbf89f66c13aab0988976adc8ff72)